### PR TITLE
CAM: Avoid bosses when pocketing.

### PR DESCRIPTION
**Pull request by sliptonic**
Prior to this fix, the 'use outline' feature would force the pocket clearing algorithm to ignore all internal wires of the selected face.  This includes holes and raised protrusions (bosses).

There is no valid case where ignoring the boss makes sense.  This PR selectively removes wires that represent 'holes' in the target face while preserving wires that represent bosses.

avoidboss

This is a renamed .FCStd, not a zipped .FCStd
[bosstest.zip](https://github.com/user-attachments/files/23607157/bosstest.zip)

### CODE REVIEW

**Comment by jffmichi** [comment]
Then we should probably rename the option to something like "ignore holes"

**Comment by sliptonic** [comment]
> Then we should probably rename the option to something like "ignore holes"

Changed the label

**Comment by tarman3** [comment]
But with this option ignores not only holes, but also extra material (do not know the correct word)
Screenshot_20251020_183515_lossy

**Comment by sliptonic** [comment]
> But with this option ignores not only holes, but also extra material (do not know the correct word) 

I don't understand your point.  With the transparency set on that image, I'm not sure what I'm looking at.

The point of the PR is that we should NEVER ignore material that protrudes upward from the face being pocketed.  A feature like the '+'  in the center of the model in the description of this PR is sometimes called a "boss" and as part of the model, it should never be ignored. To do so would damage the model.

With this change, we only ignore 'holes' with the expectation that they will be pocketed,helical, or drilled in a later operation.  I changed the label in response to @jffmichi's comment but I did not change the property name.

**Comment by tarman3** [comment]
Now it is clear

**Comment by sliptonic** [comment]
Should 'ignore holes' be enabled by default?

**Comment by tarman3** [comment]
My opinion, Yes
But probably users who do not love changes behavior will answer opposite

**Comment by jffmichi** [comment]
> Should 'ignore holes' be enabled by default?

This will not change existing files I pressume? Are you reasonably sure that this does not remove material where the model is?

Then yes 👍

**Comment by jffmichi** [comment]
Found one:

grafik

[pocket_avoid_boss_issue.zip](https://github.com/user-attachments/files/23007355/pocket_avoid_boss_issue.zip)

**Comment by sliptonic** [comment]
> Found one:

Interesting.  Good edge case!
The same problem occurs with the old implementation.  It also doesn't avoid the raised section so this PR doesn't make it any worse.  Unchecking 'ignore holes' confines the path to the outer perimeter so it is usable. 

the algorithm I'm using to detect 'bosses' is this:
1) Take the input face and build a list of wires other than the outer wire.  This list represents holes and bosses.
2) Iterate the list.  For each wire, find all the faces in the model that share edges.  These faces are either holes (down) or walls (up)
3) compare the highest point in the suspect face to the original target face.  If higher, it's a wall so keep the wire.  If lower, it's a hole so drop it.
4) Build a new face from the outerwire and the retained wires.  This is the input face to the clearing area.

in this case, the algorithm fails because the 'walls' don't share any edges with the original target face.
Any suggestions to improve it?

**Comment by jffmichi** [comment]
Another funny one:

grafik

I guess if we really want to handle all those edge cases we need to test all faces that intersect with the 2d projection of the hole outline. But at that point this is no longer a 2d operation and there is more and more overlap with adaptive and/or 3d pocket...

**Comment by tarman3** [comment]
at least can use `BoundBox.ZMax` to check if outline feature will work correct

**Comment by Dimitris75** [comment]
Use the outer wire to build a face. Slice the model on final depth. Cut the face with what left of the model?

**Comment by sliptonic** [comment]
I've implemented a new algorithm. 
Here's my test piece 
image

- holes
- holes with islands
- actual bosses
- hole with adjoined protrusion.

Not ignoring holes
image


Ignoring holes
image


Could use some more testing and weird edge cases.

**Comment by tarman3** [comment]
Great improvements
Please add your test example to top post

**Comment by sliptonic** [comment]
> Great improvements Please add your test example to top post

done

**Comment by sliptonic** [comment]
@tarman3 
I've rebased this to resolve a merge conflict with your bspline surface changes.  I think it's right.  
Can you take a look?

**Comment by tarman3** [comment]
Looks like you revert back some changes in function `classifySub()`,
but probably is enough to replace all `self.horiz.append(disk)` by `self.horiz.append((disk, bs))`

**Comment by tarman3** [comment]
Get errors if scaled model

```
 File "/home/user/projects/FreeCAD_build/Mod/CAM/Path/Op/PocketShape.py", line 189, in areaOpShapes
    horiz = [self.removeHoles(base, face) for (face, base) in self.horiz]
<class 'TypeError'>: cannot unpack non-iterable Part.Face object
20:30:29  Pocket_Shape: cannot unpack non-iterable Part.Face object
```

File: src/Mod/CAM/Path/Op/PocketShape.py [hunk 1]
```diff
@@ -198,46 +273,49 @@ def isVerticalExtrusionFace(self, face):
...
                 self.vert.append(face)
                 return True
+            else:
+                return False
+
+        elif isinstance(face.Surface, Part.BSplineSurface):
+            Path.Log.debug("face Part.BSplineSurface")
+            if Path.Geom.isRoughly(face.BoundBox.ZLength, 0):
+                Path.Log.debug("  flat horizontal or almost flat horizontal")
+                self.horiz.append(face)
```
Comment (tarman3) [inline src/Mod/CAM/Path/Op/PocketShape.py]: Probably error in line 302
`face` -> `(face, bs)`

**Comment by sliptonic** [comment]
> Get errors if scaled model
> 

Do you have a sample file?

**Comment by tarman3** [comment]
[10585.zip](https://github.com/user-attachments/files/24018203/10585.zip)

**Comment by tarman3** [comment]
Please check `Gui/Base.py`
Looks like this changes from another pull request

Probably should not here commit `suppress rendering of first rapids`

**Comment by sliptonic** [comment]
> Please check `Gui/Base.py` Looks like this changes from another pull request
> 
> Probably should not here commit `suppress rendering of first rapids`

good catch.  Thanks.
